### DISCUSSION TITLE
How to get HTTP response code alongside the returned object

### DISCUSSION THREAD

**Discussion by maroonedsia** [reactiveui/refit#1644] [Q&A] [answered]
Hi, imagine I have this interface:

```csharp
public interface IStudentEndpoints
{
	[Get("")]
	Task<List<Student>> GetAll(...);
}
```

How can I retrieve the http status code alongside the returned object, particularly when something goes wrong? I don't want the raw response; I would like to receive something like:

Happy path: The actual list of students + 200
Unhappy path 1: null + 500
Unhappy path 2: null + 404

I own the Web APIs service that the Refit client is trying to reach. So, I could create an augmented response model, for example something like:
```csharp
public class MyResponse
{
	public List<Student> Students {get;set;}
    public int HttpResponseCode {get;set;}
}
```

But, is there a built-in way to handle this gracefully by Refit? Thanks a lot! 👍

**Comment by glennawatson** [comment, marked as answer]
Yes, Refit has this built in. Change the return type to ApiResponse<T> and you get the deserialized object plus the full response metadata, without dealing with the raw stream.

```csharp
public interface IStudentEndpoints
{
    [Get("")]
    Task<ApiResponse<List<Student>>> GetAll(...);
}
```

Then:

```csharp
var resp = await api.GetAll(...);
HttpStatusCode? code = resp.StatusCode;   // 200, 404, 500, ...
bool ok = resp.IsSuccessStatusCode;
List<Student>? students = resp.Content;   // null on failure
ApiException? error = resp.Error;         // populated on failure
```

Notes:
- On a non-success status, Content is null and Error holds the ApiException (status code, reason, response body).
- ApiResponse<T> is IDisposable; dispose it (or use it in a using) since it holds the HttpResponseMessage.
- If you only want the success object and want failures to throw, keep Task<List<Student>> and catch ApiException, which also carries StatusCode.

So you do not need a hand-rolled MyResponse wrapper.
